keps/sig-node/20190806-memory-manager.md

```diff
@@ -0,0 +1,264 @@
+  Delete(containerID string)
+  ClearState()
+}
+
+type Manager interface {
+  Start(ActivePodsFunc, status.PodStatusProvider, runtimeService)
+  AddContainer(p *v1.Pod, c *v1.Container, containerID string) error
+  RemoveContainer(containerID string) error
+  State() state.Reader
+  GetTopologyHints(pod v1.Pod, container v1.Container) map[string][]topologymanager.TopologyHint
+}
+
+type Policy interface {
+  Name() string
+  Start(s state.State)
+  AddContainer(s state.State, pod *v1.Pod, container *v1.Container, containerID string) error
+  RemoveContainer(s state.State, containerID string) error
+}
+```
+
+_Listing: Memory Manager and related interfaces (sketch)._
+
+![memory-manager-interfaces](https://raw.githubusercontent.com/bg-chun/kep-draft/master/class_diagram.png)
+
+_Figure: Memory Manager components._
+
+![memory-manager-class-diagram](https://raw.githubusercontent.com/bg-chun/kep-draft/master/related_interfaces.png)
+
+_Figure: Resource Managers with topology manager_
+
+### Topology Manager changes
+
+Topology Manager takes topology hints(bits that represents NUMA node) from resource managers such as cpu manager, device manager. Then, it calculates best affinity under given policy(preferred or strict) to determine pod admission. Likewise other resource managers, Memory Manager provide hints to Topology Manager.
+
+Memory Manager implements 'Manager' interface, so that Topology Manager be able to add Memory Manager as hint provider at initialization sequence.
+
+```go
+package cm
+
+func NewContainerManager(...) (ContainerManager, error) {
+  ...
+  cm.topologyManager.AddHintProvider(cm.memoryManager)
+  ...
+}
+```
+
+### Internal Container Lifecycle changes
+
+InternalContainerLifecycle interface defines following interfaces to manage container resources, PreStartContainer, PreStopContainer, PostStopContainer. In this cases, Memory Manager has to involve managing memory/hugepages when start/stop container.
+
+Memory Manager manage memory with pod, container ID so these function needs that too.
+
+```go
+package cm
+
+func (i *internalContainerLifecycleImpl) PreStartContainer(...) {
+  ...
+  err := i.memoryManager.AddContainer(pod, containerID)
+  ...
+}
+
+func (i *internalContainerLifecycleImpl) PreStopContainer(...) {
+  ...
+  err := i.memoryManager.RemoveContainer(pod, containerID)
+  ...
+}
+
+func (i *internalContainerLifecycleImpl) PostStopContainer(...) {
+  ...
+  err := i.memoryManager.RemoveContainer(pod, containerID)
+  ...
+}
+```
+### Feature Gate and Kubelet Flags
+
+A new feature gate will be added to enable the Memory Manager feature. This feature gate will be enabled in Kubelet, and will be disabled by default in the Alpha release.
+
+- Proposed Feature Gate:  
+  `--feature-gate=MemoryManager=true`
+
+This will be also followed by a Kubelet Flag for the Memory Manager Policy, which is described above. The `none` policy will be the default policy.
+
+- Proposed Policy Flag:  
```

Comment:
why do we need a separate policy flag just for the memory manager?
the decision whether to admit / deny the Pod is made in the Topology Manager, nut in the hint providers. and we already plan to add the policy flag in the scope of the Topology Manager feature if I'm not mistaken
strict allocation policy should be per node IMO, not per resource.

Reply:
+1, if you keep this option please describe what control it provides beyond what you get with the topology manager policy

Reply:
For what purpose will we use this flag? I thought the memory manager will just inform the topology manager regarding memory and huge pages availability on each NUMA node, and according to this the topology and to the policy it will decide what NUMA node to use.